text/0014-split-contract-execution.md

```diff
@@ -0,0 +1,69 @@
+- Proposal Code Name: `split_contract_execution`
+- Start Date: 2019-09-13
+- NEP PR: [nearprotocol/neps#00014](https://github.com/nearprotocol/NEPs/pull/14)
+
+# Summary
+[summary]: #summary
+
+Split contract execution is a convenience tool, a framework built on the top of near-bindgen that allows performing
+very long and heavy computations that otherwise would not be possible to perform with a single function call because
+either they take too much time/gas or require storing too much state.
+
+For example, we want to do BLS signature aggregation of a very long list of signatures. Specifically, if we
+attempt such execution the contract will exceed the maximum gas allowed to be burnt per contract function call.
+Also, even if we introduce a host-side function (or precompile) that does aggregation we cannot directly feed it
+BLS signatures because the computation will slowdown the block production significantly.
+
+If such computation can be split into logical steps where at the end of step X we save the state and then start from it
+at step X+1 then we can split such execution across multiple contract calls. In the above example we can add one BLS
+signature to the aggregate at each step. When smart contract finishes aggregating signature X, it saves the current
+aggregate into its state and calls itself using promise API to aggregate the next signature X+1. When it reaches the
+final signature it stops calling itself.
+
+Similarly, when contract execution requires too much space we can partition it across multiple contracts (in expectation
+that they will end up on different shards). For instance ethash computation requires storing a very large graph (~1GiB)
+which we can split across multiple contract instances that perform their part of verification in parallel before calling
+the "aggregator" to call the result, i.e. MapReduce.
+
+## Non-distributed execution
+
+For non-distributed execution we can provide nice convenience functions to the contract developer using near-bindgen.
+
+Here is an example of BLS aggregator:
```

Comment:
I assume this is for few, large calculations (or batches of), vs lots of small ones? As this much provide some overhead for each async call?